File: 2024-shenzhen-FinTechathon/SpotOn/backend/check/Pedersen/verify.js
/**
 * File Name: verify.js
 * Project Name: ZeroID
 * Author(s): Norton
 * Creation Date: 2024-02-01
 * Copyright: Copyright (C) 2024 Example Corporation. All rights reserved.
 * License: This code is distributed under the MIT license.
 * Modification History:
 *    - 2024-02-02: Initial version by Norton.
 *    - 2024-02-10: Minor bug fixes by Norton.
 * Description: This is an implementation of Pedersen's Commitment in the multiplicative group, written
 *              in JavaScript language.
 * Contact Information: [None]
 */

import * as fs from 'fs'
import CryptoJS from 'crypto-js'
import { Pedersen } from './Pedersen.js'
import { Homomorphic } from './Homomorphic.js'

// 生物特征编码的维度
const dim = 128
// 承诺的生成元
const g = 56544564n
const h = 237684576n
// 假设所有变量都已定义并赋值
console.log("G: " + g);
console.log("h: " + h);

// 读取数据
try {
    // 同步读取文件内容
    //const data = fs.readFileSync(path.resolve('D:/java/code/check/Pedersen/identity.json'), 'utf8')
    const data_user = fs.readFileSync('identity.json', 'utf8')
    const data_proof = fs.readFileSync('proof.json', 'utf8')
    // 将字符串转换为JSON对象
    var json_user = JSON.parse(data_user)
    var json_proof = JSON.parse(data_proof)
} catch (err) {
    console.error("An error occurred while reading the JSON File.", err)
}

// 开始验证
console.time('time')
let pass = 1
// 获取用户身份信息
const id = json_proof.id
const t = json_proof.time
const cmt_old_json = json_user.cmt//有点问题
const cmt_new_json = json_proof.proof.emb
const cmt_eq_old_json = json_proof.proof.cmt.cmt_eq_old_json
const cmt_eq_new_json = json_proof.proof.cmt.cmt_eq_new_json
const cmt_mult_json = json_proof.proof.cmt.cmt_mult_json
const a_json = json_proof.proof.a
const z_json = json_proof.proof.z
const p_json = json_proof.proof.p



console.log("ID: " + id);
console.log("Time: " + t); // 使用JavaScript变量名
console.log("cmtOldJson: ", cmt_old_json); // 使用JavaScript变量名
console.log("cmtNewJson: ", cmt_new_json); // 使用JavaScript变量名
console.log("cmtEqOldJson: ", cmt_eq_old_json); // 使用JavaScript变量名
console.log("cmtEqNewJson: ", cmt_eq_new_json); // 使用JavaScript变量名
console.log("cmtMultJson: ", cmt_mult_json); // 使用JavaScript变量名
console.log("aJson: ", a_json); // 使用JavaScript变量名
console.log("zJson: ", z_json); // 使用JavaScript变量名
// 打印 z_json 的长度

console.log("pJson: ", p_json); // 使用JavaScript变量名
// 时间有效性判断
var t_end = new Date()
var t_start = new Date(t)
// 打印时间信息
console.log("Start Time: " + t_start);
console.log("End Time: " + t_end);
console.log("Time Difference (ms): " + (t_end - t_start));
if(t_end - t_start <= 3600000000) {
    pass = 1 * pass
}
else {
    pass = 0 * pass
}
// 自行生成挑战e
const challenge = {cmt_old_json, cmt_new_json, cmt_eq_old_json, cmt_eq_new_json, cmt_mult_json, id, t}
// e -> hash

const test = JSON.stringify(challenge)
const hash = CryptoJS.SHA256(test.toString()).toString()
const e = BigInt('0x' + hash) % 5000n
// 打印挑战信息
console.log("Challenge: ", challenge);
console.log("e: " + e);
//console.log(e)
// 数据整理
let cmt_old = new Array(dim)
let cmt_new = new Array(dim)
let cmt_eq_old = new Array(dim)
let cmt_eq_new = new Array(dim)
let cmt_mult = new Array(dim)
let z1 = new Array(dim)
let z2 = new Array(dim)
let z3 = new Array(dim)
let z4 = new Array(dim)
let z5 = new Array(dim)
let z6 = new Array(dim)
let z7 = new Array(dim)
let p1 = new Array(dim)
let p2 = new Array(dim)
let p3 = new Array(dim)
const a1 = BigInt(a_json[0])
const a2 = BigInt(a_json[1])

for(let i = 0; i < dim; i++) {
    cmt_old[i] = BigInt(cmt_old_json[i])
    cmt_new[i] = BigInt(cmt_new_json[i])
    cmt_eq_old[i] = BigInt(cmt_eq_old_json[i])
    cmt_eq_new[i] = BigInt(cmt_eq_new_json[i])
    cmt_mult[i] = BigInt(cmt_mult_json[i])
    z1[i] = BigInt(z_json[i])  
    z2[i] = BigInt(z_json[dim + i])
    z3[i] = BigInt(z_json[2 * dim + i])
    z4[i] = BigInt(z_json[3 * dim + i])
    z5[i] = BigInt(z_json[4 * dim + i])
    z6[i] = BigInt(z_json[5 * dim + i])
    z7[i] = BigInt(z_json[6 * dim + i])
    p1[i] = BigInt(p_json[i])
    p2[i] = BigInt(p_json[dim + i])
    p3[i] = BigInt(p_json[2 * dim + i])
    //console.log(emb_old[i], rand_old[i], cmt_old[i])
}

console.log("a1 :", a1);
console.log("a2: ", a2);

// 打印BigInteger数组
console.log("cmtOld: ", cmt_old);
console.log("cmtNew: ", cmt_new);
console.log("cmtEqOld: ", cmt_eq_old);
console.log("cmtEqNew: ", cmt_eq_new);
console.log("cmtMult: ", cmt_mult);
console.log("z1: ", z1);
console.log("z2: ", z2);
console.log("z3: ", z3);
console.log("z4: ", z4);
console.log("z5: ", z5);
console.log("z6: ", z6);
console.log("z7: ", z7);
console.log("p1: ", p1);
console.log("p2: ", p2);
console.log("p3: ", p3);


// 验证乘法同态的准确性
const code = new Pedersen(g, h)
let d1 = new Array(dim)
let d2 = new Array(dim)
let d3 = new Array(dim)
let d4 = new Array(dim)
let d5 = new Array(dim)
for(let i = 0; i < dim; i++) {
    d1[i] = code.Commitment(z1[i], z2[i])
    let f = new Pedersen(cmt_old[i], h)
    d2[i] = f.Commitment(z1[i], z5[i])
    d3[i] = code.Commitment(z3[i], z4[i])
    let ff = new Pedersen(cmt_new[i], h)
    d4[i] = ff.Commitment(z3[i], z6[i])
    d5[i] = f.Commitment(z3[i], z7[i])
}


// 打印验证结果数组
console.log("d1: ", d1);
console.log("d2: ", d2);
console.log("d3: ", d3);
console.log("d4: ", d4);
console.log("d5: ", d5);

// 注册用生物特征编码的平方项的准确性验证
const hp = new Homomorphic()
for(let i = 0; i < dim; i++) {
    if(d1[i] == hp.Add(a1, hp.Pow(cmt_old[i], e)) && d2[i] == hp.Add(p1[i], hp.Pow(cmt_eq_old[i], e))) {
        pass = pass * 1
    }
    else {
        pass = pass * 0
    }
}
//console.log("m->m^2: ", pass == 1)
// 验证用生物特征编码的平方项的准确性验证
for(let i = 0; i < dim; i++) {
    if(d3[i] == hp.Add(a2, hp.Pow(cmt_new[i], e)) && d4[i] == hp.Add(p2[i], hp.Pow(cmt_eq_new[i], e))) {
        pass = pass * 1
    }
    else {
        pass = pass * 0
    }
}
//console.log("m‘->m’^2: ", pass == 1)
// 不同生物特征编码的乘积项的准确性验证
for(let i = 0; i < dim; i++) {
    if(d5[i] == hp.Add(p3[i], hp.Pow(cmt_mult[i], e))) {
        pass = pass * 1
    }
    else {
        pass = pass * 0
    }
}
//console.log("mm‘->mm’^2: ", pass == 1)
// 自行计算生物特征间欧式距离的承诺
let sum_old = 1n
let sum_new = 1n
let sum_mult = 1n
for(let i = 0; i < dim; i++) {
    sum_old = hp.Add(sum_old, cmt_eq_old[i])
    sum_new = hp.Add(sum_new, cmt_eq_new[i])
    sum_mult = hp.Add(sum_mult, cmt_mult[i])
}


// 打印欧式距离的承诺
console.log("sumOld: " + sum_old);
console.log("sumNew: " + sum_new);
console.log("sumMult: " + sum_mult);
//console.log(sum_old, sum_new, sum_mult)
const cmt_sum = hp.Add(hp.Add(sum_old, sum_new), hp.Pow(hp.Add(sum_mult, sum_mult), 67108857))
if(pass == 1) {
    console.log('Verify: Pass')
}
else {
    console.log('Verify: Flase')
}
console.timeEnd('time')

File: 2024-shenzhen-FinTechathon/SpotOn/backend/check/src/main/java/szu/blockchain/check/utils/Verify.java
package szu.blockchain.check.utils;

import com.fasterxml.jackson.databind.JsonNode;
import szu.blockchain.check.entity.ProofData;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.commons.codec.digest.DigestUtils;
import szu.blockchain.check.entity.ProofData;
import java.io.File;
import java.io.IOException;
import java.math.BigInteger;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

import szu.blockchain.check.entity.Pedersen;

public class Verify {
    // 生物特征编码的维度
    private static final int DIM = 128;
    // 承诺的生成元
    private static final BigInteger g = new BigInteger("56544564");

    private static final BigInteger h = new BigInteger("237684576");
    public  boolean verify(ProofData proofData) {
        int pass = 1;
        System.out.println("G:" + g);
        System.out.println("h:" + h);



        // 获取用户身份信息
        String id = proofData.getId();
        String time = proofData.getTime();
        List<String> cmtOldJson = proofData.getOldcmt(); // 有点问题

        List<String> cmtNewJson = proofData.getProof().getEmb();
        List<String> cmtEqOldJson = proofData.getProof().getCmt().getCmtEqOldJson();
        List<String> cmtEqNewJson = proofData.getProof().getCmt().getCmtEqNewJson();
        List<String> cmtMultJson = proofData.getProof().getCmt().getCmtMultJson();
        List<String> aJson = proofData.getProof().getA();

        List<String> zJson = proofData.getProof().getZ();
        List<String> pJson = proofData.getProof().getP();

        // 打印数据
        System.out.println("ID: " + id);
        System.out.println("Time: " + time);
        System.out.println("cmtOldJson: " + cmtOldJson);
        System.out.println("cmtNewJson: " + cmtNewJson);
        System.out.println("cmtEqOldJson: " + cmtEqOldJson);
        System.out.println("cmtEqNewJson: " + cmtEqNewJson);
        System.out.println("cmtMultJson: " + cmtMultJson);
        System.out.println("aJson: " + aJson);
        System.out.println("zJson: " + zJson);
        System.out.println("pJson: " + pJson);


        // 将字符串时间转换为Date对象
        Date tStart = new Date(time);
        Date tEnd = new Date();
        System.out.println("Start Time: " + tStart);
        System.out.println("End Time: " + tEnd);
        System.out.println("Time Difference (ms): " + (tEnd.getTime() - tStart.getTime()));

        // 计算时间差
        long timeDifference = tEnd.getTime() - tStart.getTime();

        // 时间有效性判断

        if (timeDifference <= 3600000000L) {
            pass = pass;
        } else {
            pass = 0;
        }


        // 自行生成挑战e
        Map<String, Object> challenge = new LinkedHashMap<>();
        challenge.put("cmt_old_json", cmtOldJson);
        challenge.put("cmt_new_json", cmtNewJson);
        challenge.put("cmt_eq_old_json", cmtEqOldJson);
        challenge.put("cmt_eq_new_json", cmtEqNewJson);
        challenge.put("cmt_mult_json", cmtMultJson);
        challenge.put("id", id);
        challenge.put("t", time);
        System.out.println("Challenge: " + challenge);
        BigInteger e;
        try {
            ObjectMapper objectMapper = new ObjectMapper();
            String test = objectMapper.writeValueAsString(challenge);
            String hash = DigestUtils.sha256Hex(test.getBytes("UTF-8"));
            System.out.println("Hash: " + hash);
            e = new BigInteger(hash, 16).mod(BigInteger.valueOf(5000));
            System.out.println("e: " + e);
            // 这里可以继续使用e进行后续的验证逻辑
        } catch (Exception ex) {
            ex.printStackTrace();
            return false;
        }

        // 数据整理
        BigInteger[] cmtOld = new BigInteger[DIM];
        BigInteger[] cmtNew = new BigInteger[DIM];
        BigInteger[] cmtEqOld = new BigInteger[DIM];
        BigInteger[] cmtEqNew = new BigInteger[DIM];
        BigInteger[] cmtMult = new BigInteger[DIM];
        BigInteger[] z1 = new BigInteger[DIM];
        BigInteger[] z2 = new BigInteger[DIM];
        BigInteger[] z3 = new BigInteger[DIM];
        BigInteger[] z4 = new BigInteger[DIM];
        BigInteger[] z5 = new BigInteger[DIM];
        BigInteger[] z6 = new BigInteger[DIM];
        BigInteger[] z7 = new BigInteger[DIM];
        BigInteger[] p1 = new BigInteger[DIM];
        BigInteger[] p2 = new BigInteger[DIM];
        BigInteger[] p3 = new BigInteger[DIM];
//你尝试将一个以"0x"开头的十六进制字符串直接转换为BigInteger，而没有指定基数为16。为了正确解析这些十六进制字符串，你需要去掉"0x"前缀并指定基数为16。
        BigInteger a1 = new BigInteger(aJson.get(0).substring(2), 16);
        BigInteger a2 = new BigInteger(aJson.get(1).substring(2), 16);

        // 打印BigInteger值
        System.out.println("a1: " + a1);
        System.out.println("a2: " + a2);

        for (int i = 0; i < DIM; i++) {
            cmtOld[i] = new BigInteger(cmtOldJson.get(i).substring(2), 16);
            cmtNew[i] = new BigInteger(cmtNewJson.get(i).substring(2), 16);
            cmtEqOld[i] = new BigInteger(cmtEqOldJson.get(i).substring(2), 16);
            cmtEqNew[i] = new BigInteger(cmtEqNewJson.get(i).substring(2), 16);
            cmtMult[i] = new BigInteger(cmtMultJson.get(i).substring(2), 16);
            z1[i] = new BigInteger(zJson.get(i).substring(2), 16);
            z2[i] = new BigInteger(zJson.get(DIM + i).substring(2), 16);
            z3[i] = new BigInteger(zJson.get(2 * DIM + i).substring(2), 16);
            z4[i] = new BigInteger(zJson.get(3 * DIM + i).substring(2), 16);
            z5[i] = new BigInteger(zJson.get(4 * DIM + i).substring(2), 16);
            z6[i] = new BigInteger(zJson.get(5 * DIM + i).substring(2), 16);
            z7[i] = new BigInteger(zJson.get(6 * DIM + i).substring(2), 16);
            p1[i] = new BigInteger(pJson.get(i).substring(2), 16);
            p2[i] = new BigInteger(pJson.get(DIM + i).substring(2), 16);
            p3[i] = new BigInteger(pJson.get(2 * DIM + i).substring(2), 16);
        }

        // 打印整个BigInteger数组
        System.out.println("cmtOld: " + Arrays.toString(cmtOld));
        System.out.println("cmtNew: " + Arrays.toString(cmtNew));
        System.out.println("cmtEqOld: " + Arrays.toString(cmtEqOld));
        System.out.println("cmtEqNew: " + Arrays.toString(cmtEqNew));
        System.out.println("cmtMult: " + Arrays.toString(cmtMult));
        System.out.println("z1: " + Arrays.toString(z1));
        System.out.println("z2: " + Arrays.toString(z2));
        System.out.println("z3: " + Arrays.toString(z3));
        System.out.println("z4: " + Arrays.toString(z4));
        System.out.println("z5: " + Arrays.toString(z5));
        System.out.println("z6: " + Arrays.toString(z6));
        System.out.println("z7: " + Arrays.toString(z7));
        System.out.println("p1: " + Arrays.toString(p1));
        System.out.println("p2: " + Arrays.toString(p2));
        System.out.println("p3: " + Arrays.toString(p3));



        // 验证乘法同态的准确性
        Pedersen code = new Pedersen(g, h);
        BigInteger[] d1 = new BigInteger[DIM];
        BigInteger[] d2 = new BigInteger[DIM];
        BigInteger[] d3 = new BigInteger[DIM];
        BigInteger[] d4 = new BigInteger[DIM];
        BigInteger[] d5 = new BigInteger[DIM];

        for (int i = 0; i < DIM; i++) {
            d1[i] = code.commitment(z1[i], z2[i]);
            Pedersen f = new Pedersen(cmtOld[i], h);
            d2[i] = f.commitment(z1[i], z5[i]);
            d3[i] = code.commitment(z3[i], z4[i]);
            Pedersen ff = new Pedersen(cmtNew[i], h);
            d4[i] = ff.commitment(z3[i], z6[i]);
            d5[i] = f.commitment(z3[i], z7[i]);
        }

        // 打印整个BigInteger数组
        System.out.println("d1: " + Arrays.toString(d1));
        System.out.println("d2: " + Arrays.toString(d2));
        System.out.println("d3: " + Arrays.toString(d3));
        System.out.println("d4: " + Arrays.toString(d4));
        System.out.println("d5: " + Arrays.toString(d5));




        // 注册用生物特征编码的平方项的准确性验证
        Homomorphic hp = new Homomorphic();
        for (int i = 0; i < DIM; i++) {
            if (d1[i].equals(hp.Add(a1, hp.Pow(cmtOld[i], e))) && d2[i].equals(hp.Add(p1[i], hp.Pow(cmtEqOld[i], e)))) {
                pass = pass;
            } else {
                pass = 0;
            }
        }

        // 验证用生物特征编码的平方项的准确性验证
        for (int i = 0; i < DIM; i++) {
            if (d3[i].equals(hp.Add(a2, hp.Pow(cmtNew[i], e))) && d4[i].equals(hp.Add(p2[i], hp.Pow(cmtEqNew[i], e)))) {
                pass = pass;
            } else {
                pass = 0;

            }
        }

        // 不同生物特征编码的乘积项的准确性验证
        for (int i = 0; i < DIM; i++) {
            if (d5[i].equals(hp.Add(p3[i], hp.Pow(cmtMult[i], e)))) {
                pass = pass;
            } else {
                pass = 0;
            }
        }
        // 自行计算生物特征间欧式距离的承诺
        BigInteger sumOld = BigInteger.ONE;
        BigInteger sumNew = BigInteger.ONE;
        BigInteger sumMult = BigInteger.ONE;
        for (int i = 0; i < DIM; i++) {
            sumOld = hp.Add(sumOld, cmtEqOld[i]);
            sumNew = hp.Add(sumNew, cmtEqNew[i]);
            sumMult = hp.Add(sumMult, cmtMult[i]);
        }
// 打印BigInteger变量
        System.out.println("sumOld: " + sumOld);
        System.out.println("sumNew: " + sumNew);
        System.out.println("sumMult: " + sumMult);



        BigInteger cmtSum = hp.Add(hp.Add(sumOld, sumNew), hp.Pow(hp.Add(sumMult, sumMult), new BigInteger("67108857")));






        return pass==1;
    }
    //测试用

    public static void main(String[] args) {
        ObjectMapper objectMapper = new ObjectMapper();

        try {
            // 读取 proof.json 文件并转换为 ProofData 对象
            ProofData proofData = objectMapper.readValue(new File("C:\\example/proof.json"), ProofData.class);


            Verify verifier = new Verify();
            boolean result = verifier.verify(proofData);
            System.out.println("Verification result: " + result);
            // 继续处理其他数据

        } catch (IOException e) {
            e.printStackTrace();
        }
    }


}
